apps/sms/js/thread_ui.js

```diff
@@ -2512,137 +2486,16 @@ var ThreadUI = {
-      );
-    }
-
-    this.emit('recipientschange');
-  },
-
-  exactContact: function thui_exactContact(fValue) {
-    return Contacts.findExact(fValue);
-  },
-
-  searchContact: function thui_searchContact(fValue) {
-    if (!fValue) {
-      // In cases where searchContact was invoked for "input"
-      // that was actually a "delete" that removed the last
-      // character in the recipient input field,
-      // eg. type "a", then delete it.
-      return Promise.resolve(null);
-    }
-
-    return Contacts.findByString(fValue);
-  },
-
-  validateContact: function thui_validateContact(source, fValue, contacts) {
-    var isInvalid = true;
-    var index = this.recipients.length - 1;
-    var last = this.recipientsList.lastElementChild;
-    var typed = last && last.textContent.trim();
-    var isContact = false;
-    var record, length, number, contact;
-
-    if (index < 0) {
-      index = 0;
-    }
-
-    var props = ['tel'];
-    if (Settings.supportEmailRecipient) {
-      props.push('email');
-    }
-
-    // If there is greater than zero matches,
-    // process the first found contact into
-    // an accepted Recipient.
-    if (contacts && contacts.length) {
-      isInvalid = false;
-      record = contacts[0];
-      var values = props.reduce((values, prop) => {
-        var propValue = record[prop];
-        if (propValue && propValue.length) {
-          return values.concat(propValue);
-        }
-
-        return values;
-      }, []);
-      length = values.length;
-
-      // Received an exact match with a single tel or email record
-      if (source.isLookupable && !source.isQuestionable && length === 1) {
-        if (Utils.probablyMatches(values[0].value, fValue)) {
-          isContact = true;
-          number = values[0].value;
-        }
+      if (typed) {
+        Contacts.findByString(typed).then(
+          (contacts) => this.listContacts(typed, contacts)
+        );
       } else {
-        // Received an exact match that may have multiple tel records
-        for (var i = 0; i < length; i++) {
-          var propValue = values[i].value;
-          if (this.recipients.numbers.indexOf(propValue) === -1) {
-            number = propValue;
-            break;
-          }
-        }
-
-        // If number is not undefined, then it's safe to assume
-        // that this number is unique to the recipient list and
-        // can be added as an accepted recipient from the user's
-        // known contacts.
-        //
-        // It _IS_ possible for this to appear to be a duplicate
-        // of an existing accepted recipient: by display name ONLY;
-        // however this case will always have a different number.
-        //
-        if (typeof number !== 'undefined') {
-          isContact = true;
-        } else {
-          // If no number match could be made, then this
-          // contact record is actually inValid.
-          isInvalid = true;
-        }
+        this.toggleRecipientSuggestions();
```

Comment:
Looks like we call ```this.toggleRecipientSuggestions``` in all cases except when ```Contacts.findByString``` _unexpectedly_ fails, is it intentionally? If not, maybe we can remove ```this.toggleRecipientSuggestions``` from ```listContacts``` and always (when ```if (event.target.isPlaceholder)```) call it here, what do you think?